Add Wedge3D and test

osheamonn opened the pull request.

osheamonn committed:
Add Wedge3D and test

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -0,0 +1,79 @@
+ *  The shape of the inner object can be chosen to be cubical,
+ *  if the inner_sphericity is set to 0, or to a sphere, if
+ *  the inner_sphericity is set to 1. The inner_sphericity can
+ *  be set to any number between 0 and 1 for a bulged cube.
```
What is the inner radius measuring when the inner surface is not a sphere?

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+  // Wedges on z axis:
+  if (positioning_of_wedge_.axis() == Direction<3>::Axis::Zeta) {
+    return (positioning_of_wedge_.side() == Side::Upper)
+               ? std::array<std::decay_t<tt::remove_reference_wrapper_t<T>>,
```
There's a nice alias for this template parameter defined at the beginning of the function.

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+  if (positioning_of_wedge_.axis() == Direction<3>::Axis::Zeta) {
+    return (positioning_of_wedge_.side() == Side::Upper)
+               ? std::array<std::decay_t<tt::remove_reference_wrapper_t<T>>,
+                            3>{{physical_x, physical_y, physical_z}}
```
Can these be moved in, at least in some cases?  I don't know whether DataVector has a move negation operator, so the negative cases might not work.

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+  const ReturnType sy_square = square(y_equiangular_secant);
+  const ReturnType sx_square = square(x_equiangular_secant);
+  const ReturnType scaling_and_sphericity =
+      M_PI * sqrt(3.0) *
```
I think most of the sqrt(3.0) factors in this calculation cancel each other out.  (This one, for example, is used in linear combinations with other things with explicit sqrt(3.0) factors to form the matrix elements, all of which are divided by sqrt(3.0) at the end.)

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+  const ReturnType denominator_dzeta =
+      (2.0 * sqrt(sx_square + square(y_equiangular_tangent))) / sqrt(3.0);
+  const ReturnType denominator_other =
+      8.0 * pow<3>(0.5 * sqrt(3.0) * denominator_dzeta);
```
8 * (1/2)^3

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+       index_list<SpatialIndex<3, UpLo::Up, Frame::NoFrame>,
+                  SpatialIndex<3, UpLo::Lo, Frame::NoFrame>>>
+Wedge3D::inv_jacobian(const std::array<T, 3>& xi) const {
+  const auto jac = jacobian(xi);
```
Don't move const things, and here just inlining it into the next expression seems clearer anyway.

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+  return determinant_and_inverse(std::move(jac)).second;
+}
+
+void Wedge3D::pup(PUP::er& p) {
```
Missing inner_sphericity_

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+  p | positioning_of_wedge_;
+}
+
+bool operator==(const Wedge3D& lhs, const Wedge3D& rhs) noexcept {
```
Missing inner_sphericity_

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -0,0 +1,79 @@
+ public:
+  static constexpr size_t dim = 3;
+
+  explicit Wedge3D(double inner_radius, double outer_radius,
```
explicit is redundant

kidder replied:
Add documentation here of the parameters passed into the constructor

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -0,0 +1,79 @@
+  static constexpr size_t dim = 3;
+
+  explicit Wedge3D(double inner_radius, double outer_radius,
+                   Direction<3> positioning_of_wedge, double inner_sphericity);
```
I assume the constructor and member functions could be noexcept.

gsb76 replied:
change `inner_sphericity` to `sphericity`

mar-celine commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+      8.0 * pow<3>(0.5 * sqrt(3.0) * denominator_dzeta);
+  const ReturnType nonsphericity =
+      (-1.0 + inner_sphericity_) * 0.5 * denominator_dzeta;
+  const ReturnType trigonometric_nonsphericity =
```
This is the best variable name I could come up with, but I don't think it is a very good name

mar-celine commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+         "The outer radius must be larger than the inner radius.");
+  ASSERT(inner_sphericity >= 0 and inner_sphericity <= 1,
+         "Inner sphericity must be between 0 and 1");
+}
```
`ASSERT(cube_radius > 0, "The cube radius must be greater than zero.");`

`ASSERT(sphere_radius > 0, "The sphere radius must be greater than zero.");`

`ASSERT(cube_radius < sphere_radius or cube_radius > sqrt(3.0) * sphere_radius,`
         `"The boundaries of the sphere and cube must not overlap.");`

mar-celine replied:
maybe "surfaces" instead of "boundaries"

wthrowe replied:
The overlap assertion should have some sphericity adjustment, shouldn't it?

gsb76 commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -0,0 +1,79 @@
+ public:
+  static constexpr size_t dim = 3;
+
+  explicit Wedge3D(double inner_radius, double outer_radius,
```
change `inner_radius` to `cube_radius` and `outer_radius` to `sphere_radius`

fmahebert replied:
Maybe I misunderstand, but ... why? Don't we want to use these maps to make spherical shells where both the inner and outer surfaces are spherical?

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_Wedge3D.cpp:
```diff
@@ -0,0 +1,265 @@
+#include "tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.Wedge3D.Sphere",
```
Why is this test called "Sphere"?  All the maps in it have sphericity=0.

osheamonn replied:
It tests points which only lie on the spherical surface as opposed to the general surface. The test doesn't depend on the value of the sphericity.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_Wedge3D.cpp:
```diff
@@ -0,0 +1,265 @@
+  test_inverse_jacobian(map, test_point2);
+  test_inverse_jacobian(map, test_point3);
+
+  // Test that the implementation works when used with a CoordinateMap
```
This set of checks should probably be done for every map.  Let's split it out into a helper function.

gsb76 commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -0,0 +1,79 @@
+ *
+ *  \brief Three dimensional map from the cube to a wedge.
+ *
+ *  \detail The wedge is shaped such that six of them put together
```
The wedge is constructed by linearly interpolating between a face of a cube and a face of a cubed sphere, such that six wedges put together make up an object where one of the surfaces is spherical and the other surface is cubical.
The values of `cube_radius` (which is the length from the origin to a corner) and `sphere_radius` determine which surfaces are the interior and exterior. 

The `sphericity` of the spherical surface can also be adjusted to values between zero and one, where a value of 1 corresponds to a spherical surface and a value of zero corresponds to the surface of a cube. This allows for the construction of a `Domain` where one of the surfaces is that of a bulged cube.

wthrowe replied:
Also please document that the radial coordinate is the third logical coordinate.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_Wedge3D.cpp:
```diff
@@ -0,0 +1,265 @@
+  // corners of the wedge.
+  const std::array<double, 3> inner_corner{{-1.0, -1.0, -1.0}};
+  const std::array<double, 3> outer_corner{{1.0, 1.0, 1.0}};
+  const double random_inner_radius_upper_xi = inner_dis(gen);
```
This block and the next few have statements ordered up_xi, up_eta, lo_xi, lo_eta, up_zeta, lo_zeta.  Please either group by axis or up/lo.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_Wedge3D.cpp:
```diff
@@ -0,0 +1,265 @@
+  for (const auto direction : Direction<3>::all_directions()) {
+    const std::array<double, 3> delta_xi{{dxi, 0.0, 0.0}};
+    const std::array<double, 3> delta_eta{{0.0, deta, 0.0}};
+    const std::array<double, 3> delta_zeta{{0.0, 0.0, dzeta}};
```
Unused

kidder replied:
Also add one test for a case where the sphericity is 1

mar-celine commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+      make_with_value<ReturnType>(xi, 1.) /
+      sqrt(1.0 + square(x_unit_spherical_face) + square(y_unit_spherical_face));
+  const ReturnType z_inner_spherical_face = inner_radius_ * one_over_rho;
+
```
would it be easier to read if we had:
`const ReturnType rho = sqrt(1.0 + square(x_unit_spherical_face) + square(y_unit_spherical_face));`
`  const ReturnType z_inner_spherical_face = inner_radius_ *  make_with_value<ReturnType>(xi, 1.) / rho;`

mar-celine commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+      z_inner_cubical_face +
+      (z_outer_spherical_face - z_inner_cubical_face) * (zeta + 1) * 0.5;
+
+  // Assigns location of wedge based on passed "axis" and "sign" values.
```
"based on `positioning_of_wedge`"

kidder commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -0,0 +1,79 @@
+ *
+ *  \brief Three dimensional map from the cube to a wedge.
+ *
+ *  \detail The wedge is shaped such that six of them put together
```
The mapping that goes from a reference cube to a three-dimensional wedge centered on a coordinate axis covering a volume between an inner surface and outer surface.  One of the surfaces must be spherical, but the curvature of the other surface can be anything between flat (a sphericity of 0) and spherical (a sphericity of 1).

kidder commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+operator()(const std::array<T, 3>& x) const {
+  using ReturnType = std::decay_t<tt::remove_reference_wrapper_t<T>>;
+
+  const ReturnType xi = x[0];
```
references

kidder commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+Wedge3D::jacobian(const std::array<T, 3>& x) const {
+  using ReturnType = std::decay_t<tt::remove_reference_wrapper_t<T>>;
+
+  const ReturnType xi = x[0];
```
references

kidder commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+                    SpatialIndex<3, UpLo::Lo, Frame::NoFrame>>>
+      jacobian_matrix{};
+
+  auto &dxdxi = jacobian_matrix.template get<0, 0>(),
```
dx_dxi, ...

kidder commented on tests/Unit/Domain/CoordinateMaps/Test_Wedge3D.cpp:
```diff
@@ -0,0 +1,265 @@
+  const std::array<double, 3> test_point2{{1.0, 1.0, -0.5}};
+  const std::array<double, 3> test_point3{{0.7, -0.2, 0.4}};
+
+  test_forward_jacobian(map, test_point1);
```
remove forward

kidder commented on tests/Unit/Domain/CoordinateMaps/Test_Wedge3D.cpp:
```diff
@@ -0,0 +1,265 @@
+  test_forward_jacobian(map, test_point2);
+  test_forward_jacobian(map, test_point3);
+
+  test_inverse_jacobian(map, test_point1);
```
inverse -> inv

kidder commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -0,0 +1,79 @@
+ private:
+  friend bool operator==(const Wedge3D& lhs, const Wedge3D& rhs) noexcept;
+
+  double inner_radius_{};
```
I'm not sure where my comment went from yesterday:
inner_radius -> radius_of_other_surface
outer_radius -> radius_of_spherical_surface
positioning_of_wedge -> direction_of_wedge
inner_sphericity -> sphericity_of_other_surface

mar-celine commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,328 @@
+  ReturnType y_inner_cubical_face = eta * inner_radius_ / sqrt(3.0);
+  ReturnType z_inner_cubical_face =
+      make_with_value<ReturnType>(xi, inner_radius_ / sqrt(3.0));
+  const ReturnType x_unit_spherical_face = tan(M_PI_4 * xi);
```
note to self: variables of the form `tan(M_PI_4*name)` will be named `equiangular_name` and will be called so in the documentation - this name should agree with the name for the variable on line 128

osheamonn committed:
Allow make_coordinate_map to create a map from an lvalue

osheamonn committed:
Add test_coordinate_map_implementation

wthrowe commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,353 @@
+         "Inner sphericity must be between 0 and 1");
+  ASSERT(radius_of_other_surface < radius_of_spherical_surface or
+         radius_of_other_surface > sqrt(3.) * radius_of_spherical_surface *
+                                       (1 - sphericity_of_other_surface));
```
I told you to solve the equation and not try to guess the answer.
`(1 + sphericity_of_other_surface * (sqrt(3.) - 1)) * radius_of_other_surface > sqrt(3.) * radius_of_spherical_surface`

osheamonn committed:
Loosen tolerance in test_jacobian

One of the sets random values chosen to test Wedge3D did not give a numerical jacobian accurate enough to the previous tolerance.

kidder commented on tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp:
```diff
@@ -62,3 +62,37 @@ void test_inverse_jacobian(const Map& map,
 }
+
+/*!
+ * \brief Checks that the CoordinateMap 'map' functions as expected when used as
```
group and `

kidder commented on tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp:
```diff
@@ -62,3 +62,37 @@ void test_inverse_jacobian(const Map& map,
+    }
+    return point_as_tensor;
+  }();
+  CHECK(coord_map(test_point_tensor).get(0) ==
```
CAPTURE_PRECISE(random stuff) or fixed point

kidder commented on tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp:
```diff
@@ -62,3 +62,37 @@ void test_inverse_jacobian(const Map& map,
+  }();
+  CHECK(coord_map(test_point_tensor).get(0) ==
+        approx(gsl::at(map(test_point), 0)));
+  CHECK(coord_map.jacobian(test_point_tensor).get(0, 0) ==
```
check all components

kidder commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -11,55 +11,64 @@
+  ASSERT(radius_of_spherical_surface > 0,
+         "The radius of the spherical surface must be greater than zero.");
+  ASSERT(sphericity_of_other_surface >= 0 and sphericity_of_other_surface <= 1,
          "Inner sphericity must be between 0 and 1");
```
Inner sphericity -> Sphericity

kidder commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -11,55 +11,64 @@
+  const ReturnType& xi = x[0];
+  const ReturnType& eta = x[1];
+  const ReturnType& zeta = x[2];
+  ReturnType x_inner_cubical_face = xi * radius_of_other_surface_ / sqrt(3.0);
```
inner -> other

kidder commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -11,55 +11,64 @@
+      radius_of_other_surface_ * z_spherical_faces_common_factor;
 
-  // If inner_sphericity_==0, the inner_cubical_face variables correspond to the
+  // If sphericity_of_other_surface_==0, the inner_cubical_face variables
```
inner -> other

kidder commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -11,55 +11,64 @@
+  // correspond to the
   // face of a cube.
-  // If inner_sphericity_!=0, the inner_cubical_face variables have the
+  // If sphericity_of_other_surface_!=0, the inner_cubical_face variables have
```
inner -> other

kidder commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -114,129 +129,131 @@ Tensor<std::decay_t<tt::remove_reference_wrapper_t<T>>,
   }
 
-  const double scaled_inner_radius = inner_radius_ / sqrt(3.0);
+  const double scaled_radius_of_other_surface = radius_of_other_surface_;
```
?

kidder commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -18,21 +18,36 @@ namespace CoordinateMaps {
- *  if the inner_sphericity is set to 0, or to a sphere, if
- *  the inner_sphericity is set to 1. The inner_sphericity can
- *  be set to any number between 0 and 1 for a bulged cube.
+ * \detail The mapping that goes from a reference cube to a three-dimensional
```
\details

kidder commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -43,37 +58,37 @@ class Wedge3D {
-  double outer_radius_{};
-  Direction<3> positioning_of_wedge_{};
-  double inner_sphericity_{};
+  double radius_of_other_surface_{};
```
initialize to signaling NaNs

kidder commented on tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp:
```diff
@@ -82,17 +83,26 @@ void test_coordinate_map_implementation(const Map& map) {
+
+  for (size_t i = 0; i < Map::dim; ++i) {
+    for (size_t j = 0; j < Map::dim; ++j) {
+      CHECK(coord_map(test_point_tensor).get(i) ==
```
move up to i-loop

kidder reviewed (changes requested):
squash

kidder commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -0,0 +1,94 @@
+ private:
+  friend bool operator==(const Wedge3D& lhs, const Wedge3D& rhs) noexcept;
+
+  double radius_of_other_surface_{};
```
I just noticed I had an earlier comment that these should be initialized with signaling NaNs

fmahebert commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,357 @@
+  ASSERT(radius_of_spherical_surface > 0,
+         "The radius of the spherical surface must be greater than zero.");
+  ASSERT(sphericity_of_other_surface >= 0 and sphericity_of_other_surface <= 1,
+         "Inner sphericity must be between 0 and 1");
```
Inner -> Other surface's

fmahebert commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -0,0 +1,357 @@
+      (y_outer_spherical_face - y_inner_cubical_face) * (zeta + 1) * 0.5;
+  const ReturnType physical_z =
+      z_inner_cubical_face +
+      (z_outer_spherical_face - z_inner_cubical_face) * (zeta + 1) * 0.5;
```
I am confused by the variable names in the previous  ~40 lines of code.

For example, it looks like `physical_x` is obtained by interpolating in zeta between `x_inner_cubical_face` and `x_outer_spherical_surface`. But what if the inner surface is the spherical one, and the outer surface is the cubical one? What if both are spherical? What if one is of intermediate curvature?

It is not clear to me how the variables in this function are related to the "spherical surface, other surface" interface of the class. I would like to see improved names, but there should at least be comments.

osheamonn replied:
I think @gsb76 is waiting for this to be merged and is going to go through and clean up some of this and come up with clearer variable names since he's figured out some simplifications that can be made.

kidder commented:
Can you rebase (getting Will's commit from today attempting to fix the illegal exceptions)

nilsdeppe commented:
@wthrowe this cannot be merged without you approving the changes.

nilsdeppe commented:
@osheamonn this needs to be rebased now because of a different merge...
